metals/src/main/scala/scala/meta/internal/metals/DocumentSymbolProvider.scala

```diff
@@ -0,0 +1,157 @@
+package scala.meta.internal.metals
+
+import scala.meta._
+import org.eclipse.lsp4j.SymbolKind
+import org.eclipse.lsp4j.DocumentSymbol
+import MetalsEnrichments._
+import java.util.concurrent.LinkedBlockingQueue
+
+/**
+ *  Retrieves all the symbols defined in a document
+ *
+ *  If the document doesn't parse, we fall back to the latest
+ *  known snapshot of the document, if present
+ *
+ *  To avoid the snapshots to grow undefinitely, we only store
+ *  the ones for the most recently requested documents
+ */
+class DocumentSymbolProvider(buffers: Buffers) {
+
+  def empty: List[DocumentSymbol] = Nil
+
+  def documentSymbols(
+      path: AbsolutePath
+  ): List[DocumentSymbol] = {
+    path
+      .toInputFromBuffers(buffers)
+      .parse[Source]
+      .toOption
+      .map { source =>
+        val result = new SymbolTraverser(path.toString).apply(source)
+        lastDocumentSymbolResult.put((path, result))
+        result
+      }
+      .orElse(
+        lastDocumentSymbolResult.iterator.asScala
+          .collect {
+            case (p, symbols) if p == path => symbols
+          }
+          .toList
+          .headOption
+      )
+      .getOrElse(empty)
+  }
+
+  // FIXME(gabro): use a more sensible data structure
+  private var lastDocumentSymbolResult
+    : LinkedBlockingQueue[(AbsolutePath, List[DocumentSymbol])] =
+    new LinkedBlockingQueue(10)
```

Comment:
I'm abusing a `LinkedBlockingQueue` to implement a LRU cache.

I didn't want to bring in additional dependencies, but I'm thinking to implement one myself.
Wdyt @olafurpg?

(the other obvious solution is to limit ourselves to 1 snapshot at the time, but it kinda sucks if you switch between two open non-parsing buffers)

Reply:
> the other obvious solution is to limit ourselves to 1 snapshot at the time, but it kinda sucks if you switch between two open non-parsing buffers

That's true, I was gonna suggest to limit it to one file. Instead we can keep `Map[Path, DocumentSymbol]` and remove entries on `didClose`

Reply:
I've just pushed a commit that uses `removeEldestEntry` from `LinkedHashMap` to implement a LRU cache.

Reply:
I think it's better to remove entries based on `textDocument/didClose` notifications so we don't have to pick a magic number.

Reply:
👍 I'll implement it that way (I wanted to limit the moving parts of `DocumentSymbolProvider`, but I'm uneasy with the magic number too)

Reply:
> That's true, I was gonna suggest to limit it to one file. Instead we can keep `Map[Path, DocumentSymbol]` and remove entries on `didClose`

👍 this was also the way it worked before, IIRC